src/model.rs

```diff
@@ -787,88 +792,79 @@ impl Collections {
+                    if last_stop.stop_point_idx == first_stop.stop_point_idx {
+                        // We can discard when the stop points are identicals (see
+                        // Example 1 above).
+                        false
+                    } else {
+                        let prev_cal = self.calendars.get(&prev_vj.service_id);
+                        let next_cal = self.calendars.get(&next_vj.service_id);
+                        // for the stay in to be possible the vj should have at least one date in common
+                        prev_cal.map_or(false, |prev_cal| next_cal.map_or(false, |next_cal|prev_cal.overlaps(&next_cal))) &&
+                        // The stay in is not really possible when timing overlaps
+                        // between arrival of first vehicle journey and departure of
+                        // next vehicle journey (see Example 2 above).
+                        last_stop.departure_time <= first_stop.arrival_time
+                    }
+                }
+                _ => false,
+            }
+        };
+
+        self.vehicle_journeys
             .iter()
-            // Filtering out vehicle journeys without any stop time allowing us
-            // below to access first and last stop_times safely.
-            .filter(|(_, vj)| !vj.stop_times.is_empty())
-            .map(|(idx, _)| idx)
-            .collect();
-        // Init all vehicle journeys with:
-        // - no drop-off on first stop
-        // - no pick-up on last stop
+            .filter_map(|(vj_idx, vj)| vj.block_id.clone().map(|b| (b, (vj_idx, vj))))
+            .fold(
+                HashMap::<_, Vec<(Idx<VehicleJourney>, &VehicleJourney)>>::new(),
+                |mut map, (b, (vj_idx, vj))| {
+                    let other_block_id_vj = map.entry(b).or_insert_with(Vec::new);
+
+                    // for every vj we check if it can really be a stay in and if the last stop
+                    // is not in both vj (example 1)
+                    // Note: this is quadratic but should not be too costly since
+                    // the number of vj checked should be limited
+                    other_block_id_vj
+                        .iter()
+                        .for_each(|(other_vj_idx, other_vj)| {
+                            if can_chain_without_overlap(vj, other_vj) {
+                                allowed_first_drop_off_vj.insert(*other_vj_idx);
+                                allowed_last_pick_up_vj.insert(vj_idx);
+                            } else if can_chain_without_overlap(other_vj, vj) {
+                                allowed_first_drop_off_vj.insert(vj_idx);
+                                allowed_last_pick_up_vj.insert(*other_vj_idx);
+                            }
+                        });
+                    other_block_id_vj.push((vj_idx, vj));
+                    map
+                },
+            );
+
+        let vj_idxs: Vec<Idx<VehicleJourney>> =
```

Comment:
Can't you iterate on the &idx right away?

Reply:
nop since I need to mutate them, I don't see a better way to do this since there is no ìter_mut` for the moment 😞